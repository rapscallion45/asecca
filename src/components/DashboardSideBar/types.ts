/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
