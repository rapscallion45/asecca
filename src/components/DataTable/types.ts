/*
 * Data Table type definitions
 *
 * These type definitions are used in the applications data table component.
 * Only make changes to this file if the corresponding change is made in the
 * Data Table component(s)
 */

/**
 * Data Table Row Cell Value types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof DataTable
 *
 * @typedef DataTableRowCellValue
 */
export type DataTableRowCellValue =
  | string
  | Array<string>
  | boolean
  | null
  | undefined;

/**
 * Data Table Column types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof DataTable
 *
 * @typedef DataTableColumnType
 */
export type DataTableColumnType = 'string' | 'currency' | 'select' | 'check';

/**
 * Data Table Column properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableColumn
 * @prop {string} label - data table column name
 * @prop {string} key - data table column data identifier
 * @prop {DataTableColumnType} type - data table column type
 */
export interface IDataTableColumn {
  label: string;
  key: string;
  type: DataTableColumnType;
}

/**
 * Data Table Row properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableRow
 * @prop {string} label - data table row name
 */
export interface IDataTableRow {
  label: string;
}

/**
 * Data Table Cell Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCellValueCallback
 * @param {DataTableRowCellValue} value - updated cell value, can be null
 * @param {string} colKey - cell column data identifier
 * @param {number} rowIdx - cell row index in data table
 * @returns {void} - no return value
 */
export interface IDataTableEditCellValueCallback {
  (value: DataTableRowCellValue, colKey: string, rowIdx: number): void;
}

/**
 * Data Table Cell Value Retrieval callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableGetCellValueCallback
 * @param {number} rowIdx - cell row index in data table
 * @param {IDataTableColumn} column - cell column in data table
 * @returns {DataTableRowCellValue} - cell value, can be null or undefined
 */
export interface IDataTableGetCellValueCallback {
  (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue;
}

/**
 * Data Table Currency Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCurrencyCellValueCallback
 * @param {string | null} value - updated cell value
 * @returns {void} - no return value
 */
export interface IDataTableEditCurrencyCellValueCallback {
  (value: string | null): void;
}

/**
 * Data Table Checkbox Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @callback IDataTableEditCheckboxCellValueCallback
 * @param {boolean | null} value - updated cell boolean value
 * @returns {void} - no return value
 */
export interface IDataTableEditCheckboxCellValueCallback {
  (value: boolean | null): void;
}
