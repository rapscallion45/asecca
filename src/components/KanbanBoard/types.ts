/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
