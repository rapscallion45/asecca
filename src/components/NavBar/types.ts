/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
