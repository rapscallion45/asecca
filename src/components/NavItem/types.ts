/*
 * Nav Item type definitions
 *
 * These type definitions are used in the applications Nav Item component.
 */

/**
 * Nav Item configuration properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof NavItem
 *
 * @typedef INavItemConfig
 * @prop {string} title - nav item title text
 * @prop {string} path - nav item URL path
 * @prop {any} icon - nav item icon
 * @prop {boolean} newTab - nav item link opens in new tab
 * @prop {Array<INavItemConfig>} children - nav item child components
 */
export interface INavItemConfig {
  title: string;
  path?: string;
  icon?: any;
  newTab?: boolean;
  children?: Array<INavItemConfig>;
}
