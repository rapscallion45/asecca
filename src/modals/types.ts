/**
 * Type and interface definitions for global application modals
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 */

import { SvgIconProps } from '@mui/material';

/**
 * Modal button size options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonSizeType - can either be 'small', 'medium' or 'large'
 */
export type ModalButtonSizeType = 'small' | 'medium' | 'large';

/**
 * Modal button icon size options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonIconSizeType - can either be 'small', 'inherit' or 'large'
 */
export type ModalButtonIconSizeType = 'small' | 'inherit' | 'large';

/**
 * Modal button colour options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonColourType - can either be 'primary', 'secondary', 'inherit' or 'error'
 */
export type ModalButtonColourType =
  | 'primary'
  | 'secondary'
  | 'inherit'
  | 'error';

/**
 * Modal button variant options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonVariantType - can either be 'outlined', 'contained' or 'text'
 */
export type ModalButtonVariantType = 'outlined' | 'contained' | 'text';

/**
 * Modal normal button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef INormalButtonType
 * @prop {string} type - type of button
 * @prop {string} text - display text of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface INormalButtonType {
  type: 'normal';
  text?: string;
  icon: (props: SvgIconProps) => JSX.Element;
  size?: ModalButtonSizeType;
  style?: { [name: string]: string } | { [name: string]: number };
  variant?: ModalButtonVariantType;
  className?: string;
  color?: ModalButtonColourType;
}

/**
 * Modal round button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IRoundIconButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IRoundIconButtonType {
  type: 'round';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: 'small' | 'medium' | 'large';
  style?: { [name: string]: string } | { [name: string]: number };
  variant?: 'outlined' | 'contained' | 'text';
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal icon button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IIconButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonIconSizeType} iconSize - button icon size option
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IIconButtonType {
  type: 'icon';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: ModalButtonSizeType;
  style?: { [name: string]: string } | { [name: string]: number };
  iconSize?: 'small' | 'inherit' | 'large';
  className?: string;
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal menu button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IMenuItemButtonType
 * @prop {string} type - type of button
 * @prop {string} text - display text of button
 * @prop {string} icon - button icon element
 * @prop {string} closeMenu - callback handler for closing the modal
 * @prop {string} iconStyle - icon style options
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IMenuItemButtonType {
  type: 'menu';
  text: string;
  icon: (props: SvgIconProps) => JSX.Element;
  closeMenu: () => void;
  iconStyle?: { [name: string]: string } | { [name: string]: number };
  className?: string;
  color?: 'primary' | 'secondary' | 'inherit' | 'error';
}

/**
 * Modal menu button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IFabButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {string} text - display text of button
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IFabButtonType {
  type: 'fab';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: 'small' | 'medium' | 'large';
  variant?: 'extended' | 'circular';
  text?: string;
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal trigger button types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef TriggerButtonTypes - trigger button options type
 */
export type TriggerButtonTypes =
  | INormalButtonType
  | IRoundIconButtonType
  | IIconButtonType
  | IMenuItemButtonType
  | IFabButtonType;
