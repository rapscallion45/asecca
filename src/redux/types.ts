import { SnackbarKey } from 'notistack';
import { AlertColor } from '@mui/material';
import {
  CostsConfigCostSource,
  ICostsConfigDataPayload,
  ICostsConfigSaveDataPayload,
  IKanbanBoardColumn,
  IKanbanBoard,
} from '@/lib/api/api-types';

/**
 * Type and interface definitions for Redux actions
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 */

/**
 * Theme Type options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof ThemeReduxSlice
 *
 * @typedef ThemeType - theme can either be 'light' or 'dark' mode
 */
export type ThemeType = 'light' | 'dark';

/**
 * Fetch Costs Config args for async thunk requests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IFetchCostsConfigBySourceIdArgs
 * @prop {CostsConfigCostSource} source - costs config cost source option
 */
export interface IFetchCostsConfigBySourceIdArgs {
  source: CostsConfigCostSource;
  dataId: string | (string | null)[];
}

/**
 * Save Costs Config args for async thunk requests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef ISaveCostsConfigBySourceIdArgs
 * @prop {ICostsConfigSaveDataPayload} data - data payload to be sent
 */
export interface ISaveCostsConfigBySourceIdArgs {
  data: ICostsConfigSaveDataPayload;
}

/**
 * Costs Config edit redux action payload definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef ICostsConfigEditCostsPayload
 * @prop {string} colKey - table column key to be edited
 * @prop {number} rowIdx - table row index of the edited cost
 * @prop {string | null} value - updated column value
 */
export interface ICostsConfigEditCostsPayload {
  colKey: string;
  rowIdx: number;
  value: string | null;
}

/**
 * Costs Config state definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef ICostsConfigState
 * @prop {boolean} loading - costs config loading state
 * @prop {ICostsConfigDataPayload} data - currently loaded costs config data
 * @prop {ICostsConfigDataPayload} dataShadow - shadow copy of original data
 * @prop {string} error - current error message state of costs config
 * @prop {boolean} saving - saving state flag of costs config data
 * @prop {boolean} edited - costs config data has been edited flag
 */
export interface ICostsConfigState {
  loading: boolean;
  data: ICostsConfigDataPayload;
  dataShadow: ICostsConfigDataPayload;
  error?: string;
  saving: boolean;
  edited: boolean;
}

/**
 * Permissions definition
 *
 * List of possible permission levels
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof UserPermissionReduxSlice
 *
 * @type {Array<string>}
 */
export const permissions = [
  'Global',
  'Customer',
  'Project',
  'Collection',
] as const;

/**
 * User Permission Level definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof UserPermissionReduxSlice
 *
 * @typedef UserPermissionLevel
 */
export type UserPermissionLevel = (typeof permissions)[number];

/**
 * User Permission Level state definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IUserPermissionLevelState
 * @prop {UserPermissionLevel} level - current user permission level
 */
export interface IUserPermissionLevelState {
  level: UserPermissionLevel;
}

/**
 * Notification options definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef INotificationOptions
 * @prop {SnackbarKey} key - ID key of the notification
 * @prop {AlertColor} variant - notification type, i.e. "error", "success"
 * @prop {any} onClose - callback function invoked when notification is closed
 */
export interface INotificationOptions {
  key: SnackbarKey;
  variant: AlertColor;
  onClose?: any;
}

/**
 * Notification definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef INotification
 * @prop {string} message - notification message text
 * @prop {boolean} dismissed - has notification been dismissed by user and/or system
 * @prop {INotificationOptions} options - notification options
 */
export interface INotification {
  message: string;
  dismissed: boolean;
  options: INotificationOptions;
}

/**
 * Notification state definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef INotificationState
 * @prop {Array<INotification>} data - list of current notifications held in state
 */
export interface INotificationsState {
  data: Array<INotification>;
}

/**
 * Add Notification action payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IAddNotificationPayload
 * @prop {string} message - notification message text
 * @prop {AlertColor} variant - notification type, i.e. "error", "success"
 */
export interface IAddNotificationPayload {
  message: string;
  variant: AlertColor;
}

/**
 * Close Notification action payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef ICloseNotificationPayload
 * @prop {SnackbarKey} key - ID key of the notification
 */
export interface ICloseNotificationPayload {
  key: SnackbarKey;
}

/**
 * Remove Notification action payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IRemoveNotificationPayload
 * @prop {SnackbarKey} key - ID key of the notification
 */
export interface IRemoveNotificationPayload {
  key: SnackbarKey;
}

/**
 * Theme state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IThemeState
 * @prop {ThemeType} type - current theme type, i.e. 'light' or 'dark'
 */
export interface IThemeState {
  type: ThemeType;
}

/**
 * Kanban Board state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardState
 * @prop {Array<IKanbanBoard>} data - Kanban board datasets
 */
export interface IKanbanBoardState {
  loading: boolean;
  data: IKanbanBoard;
  dataShadow: IKanbanBoard;
  error?: string;
  saving: boolean;
  edited: boolean;
}

/**
 * Add Kanban Board action payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IAddKanbanBoardPayload
 * @prop {string} name - board name
 * @prop {Array<IKanbanBoardColumn>} newColumns - columns to be added to board
 */
export interface IAddKanbanBoardPayload {
  name: string;
  newColumns: Array<IKanbanBoardColumn>;
}

/**
 * Edit Kanban Board action payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IEditKanbanBoardPayload
 * @prop {string} name - board name
 * @prop {Array<IKanbanBoardColumn>} newColumns - columns to be updated
 */
export interface IEditKanbanBoardPayload {
  name: string;
  newColumns: Array<IKanbanBoardColumn>;
}

/**
 * Set Kanban Board active payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef ISetKanbanBoardActivePayload
 * @prop {number} index - board index to be set active
 */
export interface ISetKanbanBoardActivePayload {
  index: number;
}

/**
 * Add Kanban Board task payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IAddKanbanBoardTaskPayload
 * @prop {string} title - task title
 * @prop {string} description - task description
 * @prop {sting} status - task status
 * @prop {Array<IKanbanBoardTask>} subtasks - task's subtasks
 * @prop {number} newColIndex - index of added column
 */
export interface IAddKanbanBoardTaskPayload {
  name: string;
  status: string;
  newColIndex: number;
}

/**
 * Edit Kanban Board task payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IEditKanbanBoardTaskPayload
 * @prop {string} name - task name
 * @prop {sting} status - task status
 * @prop {number} newColIndex - index of added task column
 * @prop {number} prevColIndex - previous index of column
 * @prop {number} taskIndex - index of task
 */
export interface IEditKanbanBoardTaskPayload {
  name: string;
  status: string;
  newColIndex: number;
  prevColIndex: number;
  taskIndex: number;
}

/**
 * Drag Kanban Board task payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IDragKanbanBoardTaskPayload
 * @prop {number} colIndex - original index of column
 * @prop {number} prevColIndex - previous index of column
 * @prop {number} taskIndex - index of task
 */
export interface IDragKanbanBoardTaskPayload {
  colIndex: number;
  prevColIndex: number;
  taskIndex: number;
}

/**
 * Set Kanban Board subtask completed payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef ISetKanbanBoardSubtaskCompletedPayload
 * @prop {number} index - index of subtask
 * @prop {number} colIndex - previous index of column
 * @prop {number} taskIndex - index of parent task
 */
export interface ISetKanbanBoardSubtaskCompletedPayload {
  index: number;
  colIndex: number;
  taskIndex: number;
}

/**
 * Set Kanban Board task status payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef ISetKanbanBoardTaskStatusPayload
 * @prop {string} status - new status of task
 * @prop {number} colIndex - index of subtask
 * @prop {number} newColIndex - previous index of column
 * @prop {number} taskIndex - index of parent task
 */
export interface ISetKanbanBoardTaskStatusPayload {
  status: string;
  colIndex: number;
  newColIndex: number;
  taskIndex: number;
}

/**
 * Set Kanban Board task status payload
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IDeleteKanbanBoardTaskPayload
 * @prop {number} colIndex - index of task column
 * @prop {number} taskIndex - index of task to be deleted
 */
export interface IDeleteKanbanBoardTaskPayload {
  colIndex: number;
  taskIndex: number;
}

/**
 * Fetch Collections Kanban Board args for async thunk requests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef IFetchCollectionsKanbanBoardByProjectIdArgs
 * @prop {string | null} projectId - Collections project ID (optional)
 */
export interface IFetchCollectionsKanbanBoardByProjectIdArgs {
  projectId?: string | (string | null)[];
}
