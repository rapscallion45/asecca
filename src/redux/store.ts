import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
