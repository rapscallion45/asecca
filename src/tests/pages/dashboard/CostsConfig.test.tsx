import React from 'react';
import { Provider } from 'react-redux';
import { waitFor, render, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import createEmotionCache from '@/utils/createEmotionCache';
import ThemeConfig from '@/styles/theme/ThemeConfig';
import store from '@/redux/store';
import CostsConfigPage from '@/pages/dashboard/configurators/costs-config';
import mockRouter from 'next-router-mock';
import { UserPermissionLevel } from '@/redux/types';

/* test params */
const testCollectionUri: string =
  'dashboard/costs-config?collection=66135000015737072';
const testCollectionPermission: UserPermissionLevel = 'Collection';
const testCollectionQuery: string = '66135000015737072';
const testProjectUri: string =
  'dashboard/costs-config?project=66135000001502003';
const testProjectPermission: UserPermissionLevel = 'Project';
const testProjectQuery: string = '66135000001502003';
const testCustomerUri: string =
  'dashboard/costs-config?customer=106966000001368404';
const testCustomerPermission: UserPermissionLevel = 'Customer';
const testCustomerQuery: string = '106966000001368404';

/**
 * Costs Config Page Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 */
describe('Costs Config Page', () => {
  it('Renders correctly', async () => {
    /* Set the initial uri */
    mockRouter.push(testCollectionUri);

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <CostsConfigPage />
          </ThemeConfig>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Permission Level and Query', () => {
    it('Should render the passed permission level and query text', async () => {
      /** Arrange */
      const testCollectionText: string = `Costing Configuration - Lloyds Bank - ${testCollectionPermission} ${testCollectionQuery}`;
      const testProjectText: string = `Costing Configuration - Lloyds Bank - ${testProjectPermission} ${testProjectQuery}`;
      const testCustomerText: string = `Costing Configuration - Lloyds Bank - ${testCustomerPermission} ${testCustomerQuery}`;

      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <CostsConfigPage />
          </ThemeConfig>
        </Provider>
      );

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByText(testCollectionText)).toBeInTheDocument();
      });

      /** Act - update uri to Project Costs Config */
      act(() => {
        mockRouter.push(testProjectUri);
      });

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByText(testProjectText)).toBeInTheDocument();
      });

      /** Act - update uri to Customer Costs Config */
      act(() => {
        mockRouter.push(testCustomerUri);
      });

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByText(testCustomerText)).toBeInTheDocument();
      });
    });
  });
});
