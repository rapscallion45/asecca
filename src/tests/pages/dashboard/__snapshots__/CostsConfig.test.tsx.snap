// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Costs Config Page Renders correctly 1`] = `
[
  <div
    className="MuiBox-root css-78351c"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-1qbrrsr-MuiTypography-root"
    >
      Costing Configuration - Lloyds Bank - 
      Global
       
    </h4>
  </div>,
  <div
    className="MuiBox-root css-15edkaw"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-16mktj8-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      disabled={true}
      id=":r0:"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Save
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-16kplk2-MuiButtonBase-root-MuiButton-root"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Cancel
    </button>
  </div>,
]
`;
